### DISCUSSION TITLE
Why some model doesn't show overlap

### DISCUSSION THREAD

**Discussion by karan6181** [horovod/horovod#2778] [Q&A]
**Environment:**
1. Framework: MXNet
2. Framework version: 1.8.0
3. Horovod version: 0.21.0
4. MPI version: 4.0.3
5. CUDA version: 11.0
6. NCCL version: 2.7.8
7. Python version: 3.8.5
8. OS and version: Ubuntu 18.04
9. GCC version: 7.5.0

Hello, I have a question on why some model performs overlap and some doesn't. I ran the MXNet Mask R-CNN model and MXNet SSD model and I see different behavour in terms of allreduce arrival tensor. 

Mask RCNN training script: https://github.com/dmlc/gluon-cv/blob/master/scripts/instance/mask_rcnn/train_mask_rcnn.py
SSD training script: https://github.com/dmlc/gluon-cv/blob/master/scripts/detection/ssd/train_ssd.py

**Note:** The timeline was generated on 8 Node AWS p3dn instances. I have attached the timeline files for Mask RCNN and SSD as a ZIP.
[MX_timeline_files.zip](https://github.com/horovod/horovod/files/5905654/MX_timeline_files.zip)

If we look at the Mask RCNN timeline file, for 4th NEGOTIATE_ALLREDUCE column between 9000ms to 9600ms, the first 6 tensors started the ALLREDUCE roughly the same time, next 2 tensor started ALLREDUCE after 30ms and rest tensors (90%) started after 35ms. Meaning most of the tensors are starting ALLREDUCE operation at a same time which makes it hard to do overlap (backward computation and gradient communication) at same time and it impacts the performance since it is unable to do overlap. 

If we look at the SSD timeline file, for 4th NEGOTIATE_ALLREDUCE column between 22300ms to 22350ms, the total number of tensors are divided into buckets and each buckets contains around 19 tensors. Those 19 tensors starting allreduce at a same time and then once those 19 tensors allreduce has been finished, then another 19 tensors start the allreduce operation. This shows the indication of overlap.

- I am wondering what could be the reason that Mask R-CNN is not able to achieve the overlap?
- Is it something to do with how MXNet engine processes the tensors? I mean if MXNet engine would be sending the tensors to horovod at a same time then we would have seen no overlap in SSD too!
- Is there any settings that I need to tweak so that the Mask RCNN shows similar behavior as SSD in terms of overlap ?

**Comment by karan6181** [comment]
Tagging @tgaddair , @romerojosh , @leezu  to get your inputs on this.

**Comment by leezu** [comment]
SSD model uses `hybridize(static_alloc=True, static_shape=True)` whereas Mask RCNN uses `hybridize(static_alloc=args.static_alloc)`. Could you try using `--static-alloc` parameter when running the Mask RCNN script and if it doesn't help also try adding `static_shape=True`?

**Comment by karan6181** [comment]
@leezu Thanks for your input. The above timelines was with `static_alloc=True` in Mask RCNN. I tried with `static_shape=True` as well but the behavior is the same. All tensors still arrives at the same time. Also, I do see decrease in throughput while using `static_shape=True` .

**Comment by zhreshold** [comment]
key differences between *-rcnn comparing with SSD:
- two stage design(feature -> proposal + sampling), funneling paralization capability for this type of model
- larger FC layers

**Comment by romerojosh** [comment]
@karan6181 I see in the Mask RCNN script, you are setting `MXNET_EXEC_BULK_EXEC_MAX_NODE_TRAIN_BWD = 25`, while you are not setting any bulking settings for SSD. The MXNet engine will only resolve dependencies and submit operations to enqueue tensors to Horovod in between bulks. In the extreme case, if the BWD bulk is set large enough, submission of all tensors to Horovod will be postponed until all backwards pass kernels are complete. If you reduce this value or remove it, does that change the overlap behavior?

**Comment by karan6181** [comment]
Sorry for the late response. I had sync'd up offline with @romerojosh and we identified that the issue is with the MXNet training script. This has been fixed here: https://github.com/dmlc/gluon-cv/pull/1609

Thank You for the help @romerojosh
